rfcs/0000-go-generate.md

```diff
@@ -0,0 +1,177 @@
+# Go Generate Buildpack
+
+## Proposal
+The language family should have a `go generate` buildpack.
+
+<!---
+{{What changes are you purposing to the overall language family?}}
+-->
+
+## Motivation
+### Use Cases
+#### `go:generate` with a go tool
+
+##### How does this usually work in Dockerfile?
+
+#### `go:generate` with a tool installed on the system
+##### How does this work in a Dockerfile?
+
+#### `go:generate` with a binary generated during project build
+##### How does this work in a Dockerfile?
+
+<!---
+{{Why are we doing this? What pain points does this resolve? What use cases
+does it support? What is the expected outcome? Use real, concrete examples to
+make your case!}}
+-->
+
+## Implementation (Optional)
+### API
+If some file in the app source contains the directive
+```
+//go:generate some-tool
+```
+then the buildpack should require:
+```
+[[requires]]
+name = "some-tool"
+
+[[or]]
+
+[[or.requires]]
+mixin = "some-tool"
+```
```

Comment:
> What is the time-horizon for this kind of feature? Do we need to wait for stackpacks to make this feasible?

I think it'd be reasonable to start on this buildpack before stackpacks land. Given [your comment about using `go generate` to run arbitrary `go get` directives](https://github.com/paketo-buildpacks/go/pull/367#discussion_r561449264), the `go generate` buildpack might in some cases be able to get away with not requiring anything. I'll think more on how that could work.

Furthermore, asking buildpack users to write buildpacks that install a binary and put it on the path feels like a relatively small lift, especially if our buildpack authoring documentation improves. It'd be nice if this was only necessary for users who a) aren't maintaining their own stack and b) can't `go get` the tool(s) they need.

Reply:
Here's an opinion: I feel it would be better if this buildpack did not have the responsibility to "require" the tools used in the `go:generate` directive. The buildpack may simply be concerned with running `go generate` and it should be the user's responsibility to provide the required tools at build time. This can do so either by `go get`, or using separate buildpacks as is already mentioned, or they could run their own installation script as part of the app like
```
//go:generate chmod +x my_tool_install.sh
//go:generate sh ./my_tool_install.sh
//go:generate ./tool
```
or even by compiling from source code using stack-supplied tools
```
//go:generate gcc -o x a.c
//go:generate ./x
```

Reply:
While this kind of hack works, its really uncommon in the community. In fact, I've never seen anyone do this. I mentioned it as an example of what is possible, not to suggest that this was a viable solution. I think our users would be really surprised to learn that this was the expected path.

Reply:
@ryanmoran Looking back at the [Stack Buildpacks RFC](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md) there's an example of a stackpack that provides a mixin and a build plan dependency: [`mixin=google-chrome-stable` and `name=chromedriver`](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md#example-chrome-and-chromedriver).`google-chrome-stable` and `chromedriver` are different dependencies, where `chromedriver` depends on `google-chrome-stable`. The stackpack provides one or both depending on conditions outlined in the example.

With the example of `graphviz` and `dot`, the relationship is slightly different, because `dot` is installed as part of `graphviz`. But the idea of a stackpack providing mixins and/or build plan dependencies may be useful.

Suppose the `go generate` buildpack parses the `generate` directives during detection and finds a `dot` invocation. It requires `dot`. A stackpack that provides `graphviz` OR `dot` will pass detection. At build, it will run before the `go generate` buildpack and install `graphviz`. During the build phase of the `go generate` buildpack, `dot` will be available on the path.

There are some wrinkles here, though:
1. `graphviz` doesn't just install the tool `dot`, it also installs `neato`, `fdp`, etc. And it's possible that a user might require a few of those for a single project. If the `go generate` buildpack requires `dot` and `neato`, in order for the `graphviz` stackpack to pass detection, it'd need to provide:
```
[[provides]]
mixin="graphviz"

[[or]]

[[or.provides]]
name="dot"

[[or]]

[[or.provides]]
name="dot"
name="neato"
```
But then what about a user who just needs `neato`? Or a user who needs `graphviz`, `dot` and `fdp`? The combinatorics of options here quickly gets out of hand. For just the 6 diagram generators installed by `graphviz`, plus the mixin itself, you'd need 128 `[[or]]`s in your build plan contribution to represent all possible subsets. A cleverer approach to providing/requiring is necessary for something that provides so many tools.

2. If a stackpack has to "announce" all of the tools it puts on the path via build plan `[[provides]]`, this complicates the implementation of an `apt` buildpack that installs an arbitrary package.

Reply:
@arjun024 In light of @ryanmoran 's comment about typical use in the `go` community, how else might users provide the tools their `go generate` directives need? In the case of the existing [go generate buildpack](https://github.com/stefanlesperance/go-generate) the users maintain their own stack AFAIK. Is there a middle ground? Maybe using the [`build-plan` buildpack](https://github.com/paketo-community/build-plan)?